### PR TITLE
Vector2, 3 and 4 with pybind11

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

## Summary

This is a draft PR tothat create wrappers for Vector2, 3 and 4 with pybind11

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by ahcorde** [comment]
I made some changes:

 - I changed a little bit the ruby/CMakeLists.txt to avoid collisions with the pybind11 target.
 - I'm getting pybind11 with `FetchContent`. we will be always compiling the python interfaces, should I need to add a CMAKE flag ?
 - I installed the SWIG version in `lib/python/swig/` to avoid collisions with the two versions

**Comment by ahcorde** [comment]
@scpeters or @chapulina this is ready for a initial review.

File: src/ruby/CMakeLists.txt [hunk 1]
```diff
@@ -39,29 +39,35 @@ if (RUBY_FOUND)
 
   # Turn on c++
   set_source_files_properties(${swig_i_files} ruby.i PROPERTIES CPLUSPLUS ON)
+  set(SWIG_RB_LIB rbmath)
+  set(SWIG_RB_LIB_OUTPUT math)
```
Comment (chapulina) [inline src/ruby/CMakeLists.txt:43]: Same here about these Ruby changes, I think they could go in a separate PR

**Review by chapulina — DISMISSED** [review]
This works seamlessly with the `ign-gazebo` PR, great work!

If we're going to switch from SWIG to pybind11, I think we need to:

- [ ] Remove the Python SWIG bindings as we port them to pybind11. We shouldn't end up maintaining both
- [ ] Make sure the pybind11 bindings are being installed correctly when we create debs
- [ ] It would be nice not to waste all the effort we did to create SWIG bindings. How much effort do you think it would be to move them to the Ruby folder? We don't need to add tests, I think that would be too much work, right?

File: src/ruby/CMakeLists.txt (hunk 1 above)
Comment (ahcorde) [reply src/ruby/CMakeLists.txt:43]: https://github.com/ignitionrobotics/ign-math/pull/285

File: src/python_pybind11/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+find_package(pybind11 2.2 QUIET)
+
+if (NOT ${pybind11_FOUND})
+  if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.11.0")
+    include(FetchContent)
+    FetchContent_Declare(
+        pybind11
+        GIT_REPOSITORY https://github.com/pybind/pybind11
+        GIT_TAG        v2.8.1
+    )
```
[the remark this answers is not in the document]
Comment (ahcorde) [reply src/python_pybind11/CMakeLists.txt:19]: Ubuntu Focal has pybind11 v2.1, which is not compatible. We require at least 2.2. The interfaces will be not build and tested in Focal, is this an issue ?

**Comment by ahcorde** [comment]
> Remove the Python SWIG bindings as we port them to pybind11. We shouldn't end up maintaining both

If I remove `Vector2.i`, `Vector3.i` and `Vector4.i` I will break the SWIG interfaces at all. I think this should be the last step.

File: src/python_pybind11/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (chapulina) [reply src/python_pybind11/CMakeLists.txt:19]: Ohh I see. We officially support `ign-math6` on both Bionic and Focal. Is `pybind11` only needed at build time or also during runtime?

What do you recommend we do here, @j-rivero ?

**Comment by ahcorde** [comment]
Windows warnning, according with `nearbyte` docs: 

> 4) A set of overloads or a function template accepting an argument of any integral type. Equivalent to (2) (the argument is cast to double).

We need to cast this to the specific type

File: src/python_pybind11/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahcorde) [reply src/python_pybind11/CMakeLists.txt:19]: ping @j-rivero

[the remark this answers is not in the document]
Comment (j-rivero) [reply src/python_pybind11/CMakeLists.txt]: sorry for the silecente. Alex and I has been looking into the problem offline, seems like [Ubuntu Focal has version 2.4.3](https://packages.ubuntu.com/source/focal/pybind11) which should be fine for the implementation as far as I can say.

[the remark this answers is not in the document]
Comment (chapulina) [reply src/python_pybind11/CMakeLists.txt]: @ahcorde , if that version works, let's disable Python bindings on Bionic and use upstream's pybind11 on Focal and Jammy. Thanks!

[the remark this answers is not in the document]
Comment (ahcorde) [reply src/python_pybind11/CMakeLists.txt]: Removed this part. Tests are passing in Focal too

**Comment by ahcorde** [comment]
@scpeters this is ready for another review

File: src/python_pybind11/src/Vector2.hh [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    .def("distance", &Class::Distance, "Calc distance to the given point")
+    .def("length",
+         &Class::Length,
+         "Returns the length (magnitude) of the vector")
+    .def("squared_length",
+         &Class::SquaredLength,
+         "Return the square of the length (magnitude) of the vector")
+    .def("normalize", &Class::Normalize, "Normalize the vector length")
+    .def("normalized", &Class::Normalized, "Return a normalized vector")
+    .def("round", &Class::Round)
```
Comment (scpeters) [inline src/python_pybind11/src/Vector2.hh:68]: nit: add a doc-string

File: src/python_pybind11/src/Vector2.hh [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    .def("rounded", &Class::Rounded, "Get a rounded version of this vector")
+    .def("set", &Class::Set, "Set the contents of the vector")
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs", &Class::Abs,
+        "Return the absolute dot product of this vector and "
+        "another vector. This is similar to the Dot function, except the "
+        "absolute value of each component of the vector is used.")
+    .def("abs_dot", &Class::AbsDot, " Get the absolute value of the vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector2.hh:78]: I think this doc-string is swapped with `Abs`

File: src/python_pybind11/src/Vector3.hh [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+         &Class::Cross,
+         "Return the cross product of this vector with another vector.")
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs", &Class::Abs,
+         "Return the absolute dot product of this vector and "
+         "another vector. This is similar to the Dot function, except the "
+         "absolute value of each component of the vector is used.")
+    .def("abs_dot", &Class::AbsDot, " Get the absolute value of the vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector3.hh:89]: I think this doc-string is swapped with `Abs`

File: src/python_pybind11/src/Vector4.hh [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    .def("rounded", &Class::Rounded, "Get a rounded version of this vector")
+    .def("set", &Class::Set, "Set the contents of the vector")
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs", &Class::Abs,
+         "Return the absolute dot product of this vector and "
+         "another vector. This is similar to the Dot function, except the "
+         "absolute value of each component of the vector is used.")
+    .def("abs_dot", &Class::AbsDot, " Get the absolute value of the vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector4.hh:85]: I think this doc-string is swapped with Abs

File: src/python_pybind11/src/Vector2.hh [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    .def(-py::self)
+    .def("equal", &Class::Equal, "Equal to operator")
+    .def("is_finite",
+         &Class::IsFinite,
+         "See if a point is finite (e.g., not nan)")
+    .def("correct", &Class::Correct, "Corrects any nan values")
+    .def("x", py::overload_cast<>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<>(&Class::Y), "Get the y value.")
+    .def("x", py::overload_cast<const T&>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<const T&>(&Class::Y), "Get the y value.")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector2.hh:117]: I think this and the one above are setters, not getters. Please adjust the doc-string

File: src/python_pybind11/src/Vector3.hh [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    .def("is_finite",
+         &Class::IsFinite,
+         "See if a point is finite (e.g., not nan)")
+    .def("correct", &Class::Correct, "Corrects any nan values")
+    .def("x", py::overload_cast<>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<>(&Class::Y), "Get the y value.")
+    .def("z", py::overload_cast<>(&Class::Z), "Get the z value.")
+    .def("x", py::overload_cast<const T&>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<const T&>(&Class::Y), "Get the y value.")
+    .def("z", py::overload_cast<const T&>(&Class::Z), "Get the z value.")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector3.hh:142]: I think this and the two above are setters, not getters. Please adjust the doc-string

File: src/python_pybind11/src/Vector4.hh [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+         "See if a point is finite (e.g., not nan)")
+    .def("correct", &Class::Correct, "Corrects any nan values")
+    .def("x", py::overload_cast<>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<>(&Class::Y), "Get the y value.")
+    .def("z", py::overload_cast<>(&Class::Z), "Get the z value.")
+    .def("w", py::overload_cast<>(&Class::W), "Get the w value.")
+    .def("x", py::overload_cast<const T&>(&Class::X), "Get the x value.")
+    .def("y", py::overload_cast<const T&>(&Class::Y), "Get the y value.")
+    .def("z", py::overload_cast<const T&>(&Class::Z), "Get the z value.")
+    .def("w", py::overload_cast<const T&>(&Class::W), "Get the w value.")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector4.hh:128]: I think this and the three above are setters, not getters. Please adjust the doc-string

File: src/python_pybind11/test/Vector3_TEST.py [hunk 10]
```diff
@@ -99,11 +101,11 @@ def test_vector3(self):
...
         self.assertEqual(Vector3d.ONE, Vector3d(1, 1, 1))
 
-        self.assertEqual(Vector3d.UNIT_X, Vector3d(1, 0, 0))
+        self.assertEqual(Vector3d.UNITX, Vector3d(1, 0, 0))
 
-        self.assertEqual(Vector3d.UNIT_Y, Vector3d(0, 1, 0))
+        self.assertEqual(Vector3d.UNITY, Vector3d(0, 1, 0))
 
-        self.assertEqual(Vector3d.UNIT_Z, Vector3d(0, 0, 1))
+        self.assertEqual(Vector3d.UNITZ, Vector3d(0, 0, 1))
```
Comment (scpeters) [inline src/python_pybind11/test/Vector3_TEST.py:108]: this seems like an API change. Is there a pythonic reason to not use `_` in a constant name?

File: src/python_pybind11/src/Vector3.hh [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+         &Class::Cross,
+         "Return the cross product of this vector with another vector.")
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs_dot", &Class::AbsDot,
+         "Return the absolute dot product of this vector and "
+         "another vector. This is similar to the Dot function, except the "
+         "absolute value of each component of the vector is used.")
+    .def("abs", &Class::Abs, " Get the absolute value of the vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector3.hh:89]: nit: leading whitespace in doc-string

File: src/python_pybind11/src/Vector4.hh [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    .def("rounded", &Class::Rounded, "Get a rounded version of this vector")
+    .def("set", &Class::Set, "Set the contents of the vector")
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs_dot", &Class::AbsDot,
+         "Return the absolute dot product of this vector and "
+         "another vector. This is similar to the Dot function, except the "
+         "absolute value of each component of the vector is used.")
+    .def("abs", &Class::Abs, " Get the absolute value of the vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector4.hh:85]: nit: leading whitespace in doc-string

File: src/python_pybind11/src/Vector2.hh [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    .def("dot",
+         &Class::Dot,
+         "Return the dot product of this vector and another vector")
+    .def("abs_dot", &Class::AbsDot,
+        "Return the absolute dot product of this vector and "
+        "another vector. This is similar to the Dot function, except the "
+        "absolute value of each component of the vector is used.")
+    .def("abs",
+         &Class::Abs,
+         "Return the absolute dot product of this vector and another vector")
```
Comment (scpeters) [inline src/python_pybind11/src/Vector2.hh:82]: replace with "Get the absolute value of the vector"

File: src/python_pybind11/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+  else()
+    # If not a system installation, respect local paths
+    set(IGN_PYTHON_INSTALL_PATH ${IGN_LIB_INSTALL_DIR}/python)
+  endif()
+
+  set(IGN_PYTHON_INSTALL_PATH "${IGN_PYTHON_INSTALL_PATH}/ignition")
+
+  # Set the build location and install location for a CPython extension
+  function(configure_build_install_location _library_name)
+    # Install into test_rclpy folder in build space for unit tests to import
```
Comment (scpeters) [inline src/python_pybind11/CMakeLists.txt:48]: nit: `test_rclpy`

**Review by scpeters — COMMENTED** [review]
just some minor comments related to comments and doc-strings, looks good!

**Review by scpeters — APPROVED** [review]
looks great!

please edit the commit message when squash+merging, thanks!

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1
